# Chapter 9: What we have covered so far (and a bit more)

---

In this chapter, we will work our way through a concise review of the Python functionality we have covered so far. Throughout this chapter, we will work with a interesting, yet not too large dataset, namely the well-known *Arabian nights*. *Alf Laylah Wa Laylah*, *the Stories of One Thousand and One Nights* is a collection of folk tales, collected over many centuries by various authors, translators, and scholars across West, Central and South Asia and North Africa, forms a huge narrative wheel with an overarching plot, created by the frame story of Shahrazad.

The stories begin with the tale of king Shahryar and his brother, who, having both been deceived by their respective Sultanas, leave their kingdom, only to return when they have found someone who — in their view — was wronged even more. On their journey the two brothers encounter a huge jinn who carries a glass box containing a beautiful young woman. The two brothers hide as quickly as they can in a tree. The jinn lays his head on the girl’s lap and as soon as he is asleep, the girl demands the two kings to make love to her or else she will wake her ‘husband’. They reluctantly give in and the brothers soon discover that the girl has already betrayed the jinn ninety-eight times before. This exemplar of lust and treachery strengthens the Sultan’s opinion that all women are wicked and not to be trusted. 

When king Shahryar returns home, his wrath against women has grown to an unprecedented level. To temper his anger, each night the king sleeps with a virgin only to execute her the next morning. In order to make an end to this cruelty and save womanhood from a "virgin scarcity", Sharazad offers herself as the next king’s bride. On the first night, Sharazad begins to tell the king a story, but she does not end it. The king’s curiosity to know how the story ends, prevents him from executing Shahrazad. The next night Shahrazad finishes her story, and begins a new one. The king, eager to know the ending of this tale as well, postpones her execution once more. Using this strategy for One Thousand and One Nights in a labyrinth of stories-within-stories-within-stories, Shahrazad attempts to gradually move the king’s cynical stance against women towards a politics of love and justice (see Marina Warner’s *Stranger Magic* (2013) in case you're interested).

The first European version of the Nights was translated into French by Antoine Galland. Many translations (in different languages) followed, such as the (heavily criticized) English translation by Sir Richard Francis Burton entitled *The Book of the Thousand and a Night* (1885). This version is freely available from the Gutenberg project (see [here](http://www.gutenberg.org)), and will be the one we will explore here.

#Files and directories

In the notebooks we use, there is a convenient way to quickly inspect the contents of a folder. Our Arabian nights are contained under the general `data``
 folder:

In [133]:
ls data/arabian_nights

1.txt     21.txt    322.txt   437.txt   55.txt    662.txt   775.txt   888.txt
10.txt    210.txt   323.txt   438.txt   550.txt   663.txt   776.txt   889.txt
100.txt   211.txt   324.txt   439.txt   551.txt   664.txt   777.txt   89.txt
1000.txt  212.txt   325.txt   44.txt    552.txt   665.txt   778.txt   890.txt
1001.txt  213.txt   326.txt   440.txt   553.txt   666.txt   779.txt   891.txt
101.txt   214.txt   327.txt   441.txt   554.txt   667.txt   78.txt    892.txt
102.txt   215.txt   328.txt   442.txt   555.txt   668.txt   780.txt   893.txt
103.txt   216.txt   329.txt   443.txt   556.txt   669.txt   781.txt   894.txt
104.txt   217.txt   33.txt    444.txt   557.txt   67.txt    782.txt   895.txt
105.txt   218.txt   330.txt   445.txt   558.txt   670.txt   783.txt   896.txt
106.txt   219.txt   331.txt   446.txt   559.txt   671.txt   784.txt   897.txt
107.txt   22.txt    332.txt   447.txt   56.txt    672.txt   785.txt   898.txt
108.txt   220.txt   333.txt   448.txt   560.txt   673.txt   786.t

As you can see, this folder holds a number of plain text files, ending in the `.txt` extension. Let us open a random file:

In [138]:
f = open('data/arabian_nights/848.txt', 'r')
text = f.read()
f.close()
print(text[:500])

,
She pursued, It hath reached me, O auspicious King, that when Zayn al-Mawasif took from the Kazi the deed which made over her lover's property to her, she said to him, "O Masrur, now gang thy gait." But her slave-girl Hubub turned to him and said, "Recite us some verses." So he improvised upon that game of chess these couplets,
"Of Time and what befel me I complain, * Mourning my loss by chess and eyes of bane.  
 For love of gentlest, softest-sided fair * Whose like is not of maids or mortal 


Here, we use the `open()` function to create a file object `f`, which we can use to access the actual text content of the file. Make sure that you do not pass the `'w'` parameter ("write") to `open()`, instead of `'r'`, since this would overwrite and thus erase the existing file.  After assigning the string returned by `f.read()` to the variable `text`, we print the 500 first characters of `text` to get an impression of what it contains, using simple string indexing (`[:500]`). Don't forget to close the file again after you have opened or strange things could happen to your file! One little trick which is commonly used to avoid having to explicitly opening and closing your file is a `with` block:

In [139]:
with open('data/arabian_nights/848.txt', 'r') as f:
    text = f.read()
print(text[:500])

,
She pursued, It hath reached me, O auspicious King, that when Zayn al-Mawasif took from the Kazi the deed which made over her lover's property to her, she said to him, "O Masrur, now gang thy gait." But her slave-girl Hubub turned to him and said, "Recite us some verses." So he improvised upon that game of chess these couplets,
"Of Time and what befel me I complain, * Mourning my loss by chess and eyes of bane.  
 For love of gentlest, softest-sided fair * Whose like is not of maids or mortal 


This code block does exactly the same thing as the previous one but saves you some typing. In this chapter we would like to work with all the files in the `arabian_nights` directory. This is where loops come in handy of course, since what we really would like to do, is iterate over the contents of the directory. Accessing these contents in Python is easy, but requires importing some extra functionality. In this case, we need to import the `os` module, which contains all functionality related to the 'operating system' of your machine, such as directory information:  

In [140]:
import os

Using the dot-syntax (`os.xxx`), we can now access all functions that come with this module, such as `listdir()`, which returns a list of the items which are included under a given directory

In [141]:
filenames = os.listdir('data/arabian_nights')
print(len(filenames))
print(filenames[:20])

999
['1.txt', '10.txt', '100.txt', '1000.txt', '1001.txt', '101.txt', '102.txt', '103.txt', '104.txt', '105.txt', '106.txt', '107.txt', '108.txt', '109.txt', '11.txt', '110.txt', '111.txt', '112.txt', '113.txt', '114.txt']


The function `os.listdir()` returns a list of strings, representing the filenames contained under a directory.

## Quiz

1. In Burton's translation some of the 1001 nights are missing. How many?
2. Can you come up with a clever way to find out which nights are missing? Hint: a counting loop and some string casting might be useful here!

In [142]:
# your code goes here

With `os.listdir()`, you need to make sure that you pass the correct path to an existing directory:  

In [143]:
os.listdir('data/belgian_nights')

FileNotFoundError: [Errno 2] No such file or directory: 'data/belgian_nights'

It might therefore be convenient to check whether a directory actually exists in a given location:

In [144]:
print(os.path.isdir('data/arabian_nights'))
print(os.path.isdir('data/belgian_nights'))

True
False


The second directory, naturally, does not exist and `isdir()` evaluates to `False` in this case. Creating a new (and thus empty) directory is also easy using `os`:

In [146]:
os.mkdir('belgian_nights')

We can see that it lives in the present working directory now, by typing `ls` agian:

In [147]:
ls

Chapter 1 - Variables.ipynb*           README.md
Chapter 2 - Collections.ipynb*         assignment/
Chapter 3 - Conditions.ipynb*          belgian_nights/
Chapter 4 - Loops.ipynb*               data/
Chapter 5 - Functions and Files.ipynb* exercise_keys/
Chapter 6 - Regular Expressions.ipynb  images/
Chapter 7 - More on Loops.ipynb        start-osx.command*
Chapter 8 - Parsing XML.ipynb          start-unix.sh*
Chapter x - Text analysis.ipynb        start-windows.bat*
LICENSE                                styles/


Or we use Python:

In [148]:
print(os.path.isdir('belgian_nights'))

True


Removing directories is also easy, but PLEASE watch out, sometimes is is too easy: if you remove a wrong directory in Python, it will be gone forever... Unlike other applications, Python does not keep a copy of it in your Trash and it does not have a Ctrl-Z button. Please watch out with what you do, since with great power comes great responsiblity! Removing the entire directory which we just created can be done as follows:  

In [149]:
import shutil
shutil.rmtree('belgian_nights')

And lo behold: the directory has disappeared again:

In [150]:
print(os.path.isdir('belgian_nights'))

False


Here, we use the `rmtree()` command to remove the entire directory in a *recursive* way: even if the directory isn't empty and contains files and subfolers, we will remove all of them. The `os` module also comes with a `rmdir()` but this will not allow you to remove a directory which is not empty, as becomes clear in the `OSError` raised below:  

In [151]:
os.rmdir('data/arabian_nights')

OSError: [Errno 66] Directory not empty: 'data/arabian_nights'

The folder contains things and therefore cannot be removed using this function. There are, of course, also ways to remove individual files or check whether they exist:

In [152]:
os.mkdir('belgian_nights')
f = open('belgian_nights/1001.txt', 'w')
f.write('Content')
f.close()
print(os.path.exists('belgian_nights/1001.txt'))
os.remove('belgian_nights/1001.txt')
print(os.path.exists('belgian_nights/1001.txt'))

True
False


Here, we create a directory, write a new file to it (`1001.txt`), and removed it again. Using `os.path.exists()` we monitored at which point the file existed. Finally, the `shutil` module also ships with a useful `copyfile()` function which allows you to copy files from one location to another, possible with another name. To copy night 66 to the present directory, for instance, we could do: 

In [153]:
shutil.copyfile('data/arabian_nights/66.txt', 'new_66.txt')

'new_66.txt'

Indeed, we have added an exact copy of night 66 to our present working directory:

In [154]:
ls

Chapter 1 - Variables.ipynb*           assignment/
Chapter 2 - Collections.ipynb*         belgian_nights/
Chapter 3 - Conditions.ipynb*          data/
Chapter 4 - Loops.ipynb*               exercise_keys/
Chapter 5 - Functions and Files.ipynb* images/
Chapter 6 - Regular Expressions.ipynb  new_66.txt
Chapter 7 - More on Loops.ipynb        start-osx.command*
Chapter 8 - Parsing XML.ipynb          start-unix.sh*
Chapter x - Text analysis.ipynb        start-windows.bat*
LICENSE                                styles/
README.md


We can safely remove it again:

In [155]:
os.remove('new_66.txt')

##Paths

The paths we have used so far are 'relative' paths, in the sense that they are relative to the place on our machine from which we execute our Python code. Absolute paths can also be retrieved and will differ on each computer, because they typically include user names etc:

In [156]:
os.path.abspath('data/arabian_nights/848.txt')

'/Users/mike/GitRepos/ghent1516/data/arabian_nights/848.txt'

While absolute paths are longer to type, they have the advantage that they can be used anywhere on your computer (i.e. irrespective of where you run your code from). Paths can be tricky. Suppose that we would like to open one of our filenames:

In [157]:
filenames = os.listdir('data/arabian_nights')
rnd_filename = filenames[9]
with open(rnd_filename, 'r') as f:
    text = f.read()
print(text[:500])

FileNotFoundError: [Errno 2] No such file or directory: '105.txt'

Python throws a `FileNotFoundError`, complaining that the file we wish to open does not exist. This situation stems from the fact that `os.listdir()` only returns the basic name of a given file, and not an entire (absolute or relative) path to it. To properly access the file, we must therefore not forget to include the rest of the path again: 

In [158]:
filenames = os.listdir('data/arabian_nights')
rnd_filename = filenames[9]
with open('data/arabian_nights/'+rnd_filename, 'r') as f:
    text = f.read()
print(text[:500])

,
She said, It hath reached me, O auspicious King, that they laid Sharrkan out and buried him in the mountain aforesaid and mourned over his far-famed virtues. Then they looked for the opening of the city gate; but it opened not and no sign of men appeared to them on the walls; whereat they wondered with exceeding wonder. But King Zau al-Makan said, "By Allah, I will not turn back from them, though I sit here for years and years, till I take blood revenge for my brother Sharrkan and waste Consta


Apart from `os.listdir()` there are a number of other common to obtain directory listings in Python. Using the `glob` module for instance, we can easily access the full relative path leading to our Arabian Nights:  

In [159]:
import glob
filenames = glob.glob('data/arabian_nights/*')
print(filenames[:10])

['data/arabian_nights/1.txt', 'data/arabian_nights/10.txt', 'data/arabian_nights/100.txt', 'data/arabian_nights/1000.txt', 'data/arabian_nights/1001.txt', 'data/arabian_nights/101.txt', 'data/arabian_nights/102.txt', 'data/arabian_nights/103.txt', 'data/arabian_nights/104.txt', 'data/arabian_nights/105.txt']


The asterisk (`*`) in the argument passed to `glob.glob()` is worth noting here. Just like with regular expressions, this asterisk is a sort of wildcard which will match any series of characters (i.e. the filenames under `arabian_nights`). When we exploit this wildcard syntax, `glob.glob()` offers another distinct advantage: we can use it to easily filter out filenames which we are not interested in:

In [160]:
filenames = glob.glob('data/arabian_nights/*.txt')
print(filenames[:10])

['data/arabian_nights/1.txt', 'data/arabian_nights/10.txt', 'data/arabian_nights/100.txt', 'data/arabian_nights/1000.txt', 'data/arabian_nights/1001.txt', 'data/arabian_nights/101.txt', 'data/arabian_nights/102.txt', 'data/arabian_nights/103.txt', 'data/arabian_nights/104.txt', 'data/arabian_nights/105.txt']


Interestingly, the command in this code block will only load filenames that end in `".txt"`. This is interesting when we would like to ignore other sort of junk files etc. that might be present in a directory. To replicate similar behaviour with `os.listdir()`, we would have needed a typical `for`-loop, such as:

In [161]:
filenames = []
for fn in os.listdir('data/arabian_nights'):
    if fn.endswith('.txt'):
        filenames.append(fn)
print(filenames[:10])

['1.txt', '10.txt', '100.txt', '1000.txt', '1001.txt', '101.txt', '102.txt', '103.txt', '104.txt', '105.txt']


Or for you stylish coders out there, with a list comprehension:

In [162]:
filenames = [fn for fn in os.listdir('data/arabian_nights') if fn.endswith('.txt')]

However, when using `glob.glob()`, you might sometimes want to be able to extract a file's base name again. There are several solutions to this:

In [163]:
filenames = glob.glob('data/arabian_nights/*.txt')
fn = filenames[10]

# simple string splittng:
print(fn.split('/')[-1])

# using os.sep:
print(fn.split(os.sep)[-1])

# using os.path:
print(os.path.basename(fn))

106.txt
106.txt
106.txt


Finally, sometimes, you might be interested in all the subdirectories of a particular directory (and all the subdirectories of these subdirectories etc.). Parsing such deep directory structures, can be tricky, especially if you do not know how deep a directory tree might run. You could of course try stacking multiple loops using `os.listdir()`, but a more convenient way is `os.walk()`:

In [164]:
for root, directory, filename in os.walk("data"):
    print(filename)

['.DS_Store', 'austen-emma-excerpt-tokenised.txt', 'austen-emma-excerpt.txt', 'austen-emma.txt', 'words.txt']
['1.txt', '10.txt', '100.txt', '1000.txt', '1001.txt', '101.txt', '102.txt', '103.txt', '104.txt', '105.txt', '106.txt', '107.txt', '108.txt', '109.txt', '11.txt', '110.txt', '111.txt', '112.txt', '113.txt', '114.txt', '115.txt', '116.txt', '117.txt', '118.txt', '119.txt', '12.txt', '120.txt', '121.txt', '122.txt', '123.txt', '124.txt', '125.txt', '126.txt', '127.txt', '128.txt', '129.txt', '13.txt', '130.txt', '131.txt', '132.txt', '133.txt', '134.txt', '135.txt', '136.txt', '137.txt', '138.txt', '139.txt', '14.txt', '140.txt', '141.txt', '142.txt', '143.txt', '144.txt', '145.txt', '146.txt', '147.txt', '148.txt', '149.txt', '15.txt', '150.txt', '151.txt', '152.txt', '153.txt', '154.txt', '155.txt', '156.txt', '157.txt', '158.txt', '159.txt', '16.txt', '160.txt', '161.txt', '162.txt', '163.txt', '164.txt', '165.txt', '166.txt', '167.txt', '168.txt', '169.txt', '17.txt', '170.t

As you can see, `os.walk()` allows you to efficiently loop over the entire tree. As always, don't forget that help is right around the corner in your notebooks. Using `help()`, you can quickly access the documentation of modules (but only after you have imported them!).

In [165]:
help(os.walk)

Help on function walk in module os:

walk(top, topdown=True, onerror=None, followlinks=False)
    Directory tree generator.
    
    For each directory in the directory tree rooted at top (including top
    itself, but excluding '.' and '..'), yields a 3-tuple
    
        dirpath, dirnames, filenames
    
    dirpath is a string, the path to the directory.  dirnames is a list of
    the names of the subdirectories in dirpath (excluding '.' and '..').
    filenames is a list of the names of the non-directory files in dirpath.
    Note that the names in the lists are just names, with no path components.
    To get a full path (which begins with top) to a file or directory in
    dirpath, do os.path.join(dirpath, name).
    
    If optional arg 'topdown' is true or not specified, the triple for a
    directory is generated before the triples for any of its subdirectories
    (directories are generated top down).  If topdown is false, the triple
    for a directory is generated after the 

## Quiz

In the next part of this chapter, we will need a way to sort our stories from the first, to the very last night. For our own convenience we will use a little hack for this. In this quiz, we would like you to create a new folder under `data` directory, called '1001'. You should copy all the original files from `arabian_nights` to this new folder, but give the files a new name, prepending zeros to filename until all nights have four digits in their name. `1001.txt` stays `1001.txt`, for instance, but `66.txt` becomes `0066.txt` and `2.txt` becomes `0002.txt` etc. This will make sorting the nights easier below. For this quiz you could for instance use a for loop in combination with a while loop (but don't get stuck in endless loops...)

In [136]:
# your code

##Parsing files

In this chapter we will introduce you to the task of text analysis in Python. You will learn how to read an entire corpus into Python, clean it and how to perform certain data analyses on those texts. We will also briefly introduce you to using Python's plotting library *matplotlib*, with which you can visualize your data.

Before we delve into the main subject of this chapter, text analysis, we will first write a couple of utility functions that build upon the things you learnt in the previous chapter. Often we don't work with a single text file stored at our computer, but with multiple text files or entire corpora. We would like to have a way to load a corpus into Python.

Remember how to read files? Each time we had to open a file, read the contents and then close the file. Since this is a series of steps we will often need to do, we can write a single function that does all that for us. We write a small utility function `read_file(filename)` that reads the specified file and simply returns all contents as a single string.

In [1]:
def read_file(filename):
    "Read the contents of FILENAME and return as a string."
    infile = open(filename) # windows users should use codecs.open after importing codecs
    contents = infile.read()
    infile.close()
    return contents

Now, instead of having to open a file, read the contents and close the file, we can just call the function `read_file` to do all that:

In [2]:
text = read_file("data/austen-emma-excerpt.txt")
print(text)

Emma by Jane Austen 1816

VOLUME I

CHAPTER I


Emma Woodhouse, handsome, clever, and rich, with a comfortable home
and happy disposition, seemed to unite some of the best blessings
of existence; and had lived nearly twenty-one years in the world
with very little to distress or vex her.

She was the youngest of the two daughters of a most affectionate,
indulgent father; and had, in consequence of her sister's marriage,
been mistress of his house from a very early period.  Her mother
had died too long ago for her to have more than an indistinct
remembrance of her caresses; and her place had been supplied
by an excellent woman as governess, who had fallen little short
of a mother in affection.


In the directory `data/gutenberg/training` we have a corpus consisting of multiple files with the extension `.txt`. This corpus is a collection of English novels which we downloaded for you from the [Gutenberg](http://www.gutenberg.org) project. We want to iterate over all these files. You can do this using the `listdir` function from the `os` module. We import this function as follows:

In [3]:
from os import listdir

After that, the `listdir` function is available to use. This function takes as argument the path to a directory and returns all the files and subdirectories present in that directory:

In [4]:
listdir("data")

['.DS_Store',
 'arabian_nights',
 'austen-emma-excerpt-tokenised.txt',
 'austen-emma-excerpt.txt',
 'austen-emma.txt',
 'british-novels',
 'gutenberg',
 'haggard',
 'names',
 'twitter.txt']

Notice that `listdir` returns a list and we can iterate over that list. Now, consider the following function:

In [5]:
def list_textfiles(directory):
    "Return a list of filenames ending in '.txt' in DIRECTORY."
    textfiles = []
    for filename in listdir(directory):
        if filename.endswith(".txt"):
            textfiles.append(directory + "/" + filename)
    return textfiles

The function `listdir` takes as argument the name of a directory and lists all filenames in that directory. We iterate over this list and append each filename that ends with the extension, `.txt` to a new list of `textfiles`. Using the `list_textfiles` function, the following code will read all text files in the directory `data/gutenberg/training` and outputs the length (in characters) of each:

In [6]:
for filepath in list_textfiles("data/gutenberg/training"):
    text = read_file(filepath)
    print(filepath +  " has " + str(len(text)) + " characters.")

data/gutenberg/training/austen-emma.txt has 887071 characters.
data/gutenberg/training/austen-pride.txt has 684765 characters.
data/gutenberg/training/austen-sense.txt has 673022 characters.
data/gutenberg/training/blake-poems.txt has 38153 characters.
data/gutenberg/training/blake-songs.txt has 32223 characters.
data/gutenberg/training/bryant-stories.txt has 243901 characters.
data/gutenberg/training/burgess-busterbrown.txt has 82992 characters.
data/gutenberg/training/carroll-alice.txt has 144395 characters.
data/gutenberg/training/chesterton-ball.txt has 457450 characters.
data/gutenberg/training/chesterton-thursday.txt has 320525 characters.
data/gutenberg/training/edgeworth-parents.txt has 916861 characters.
data/gutenberg/training/melville-piazza.txt has 467970 characters.
data/gutenberg/training/milton-paradise.txt has 468220 characters.
data/gutenberg/training/shakespeare-caesar.txt has 112310 characters.
data/gutenberg/training/shakespeare-hamlet.txt has 162881 characters.
dat

---

## Sentence tokenization

In the previous chapter we wrote a function to tokenize or split a text string into a list of words. However, using this function we lose information about where sentences end and start in the text. We will develop a function `split_sentences` that performs some very simple sentence splitting when passed a text string. Each sentence will be represented as a new string, so the function as a whole returns a list of sentence strings. We assume that any occurrence of either `.` or `!` or `?` marks the end of a sentence. In reality, this is more ambiguous of course. Consider for example the use of periods as end-of-sentence marker as well as in abbreviations and initials!

How should we tackle this problem? Have a look at the following picture:

![caption](files/images/indexing.png)

The first sentence *Hello there!* spans from index 0 to index 11. The second sentence from 13 to 26. If we come up with a way to extract those indexes, we could slice the text into separate sentences. First we define a utility function `end_of_sentence` that takes as argument a character and returns `True` if it is an end-of-sentence marker, otherwise it returns `False`. 

---

#### Quiz!

Write the function `end_of_sentence_marker` below:

In [ ]:
def end_of_sentence_marker(character):
    # insert your code here
    

# these tests should return True if your code is correct
print(end_of_sentence_marker("?") == True)
print(end_of_sentence_marker("a") == False)

---

An important function we will use is the built in `enumerate`. `enumerate` takes as argument any iterable (a string a list etc.). Let's see it in action:

In [ ]:
for element in enumerate("Python"):
    print(element)

As you can see, enumerate allows you to iterate over an iterable and for each element in that iterable, it gives you its corresponding index. A slightly more convenient way of iterating over `enumerate` is the following:

In [ ]:
for index, character in enumerate("Python"):
    print(index)

This way, we have easy access to both the index and the original item in the iterable. Now we know enough to write our `split_sentences` function. We will walk you through it, step by step, but first try to read the function and think about what it possibly does at each step:

In [ ]:
def split_sentences(text):
    "Split a text string into a list of sentences."
    sentences = []
    start = 0
    for end, character in enumerate(text):
        if end_of_sentence_marker(character):
            sentence = text[start: end + 1]
            sentences.append(sentence)
            start = end + 1
    return sentences

The function `split_sentences` takes as argument a text represented by a simple string. Within the function we define a variable `sentences` in which we will store the individual sentences. We need to extract both the start position and the end position of each sentence. We know that the first sentence will always start at position 0. Therefore we define a variable start and set it to zero.

Next we will use `enumerate` to loop over all individual characters in the text. Remember that enumerate returns pairs of indexes and their corresponding elements (here characters). For each character we check whether it is an end-of-sentence marker. If it is, the variable `end` marks the position in `text` where a sentence ends. We can now slice the text from the starting position to the end position and obtain our sentence. Notice that we add 1 to the end position. Why would that be? This is because, as you might remember from the first chapter, slices are non-inclusive, so `text[start:end]` would return the text starting at `start` and ending one position before `end`. Since we have reached the end of a sentence, we know that the next sentence will start at least one position later than our last end point. Therefore, we update the start variable to `end + 1`. Let's check whether our function works as promised:

In [ ]:
print(split_sentences("This is a sentence. Should we seperate it from this one?"))

It does! 

---

#### Quiz!

To conclude this section, you will write a wrapper function `tokenize`, that takes as input a text represented by a string and tokenizes this string into sentences. After that, we clean each sentence, by lowercasing all words and removing punctuation. The final step is to tokenize each sentence into a list of words. The file `preprocessing.py` contains a function called `clean_text` which removes all punctuation from a text and turns all characters to lowercase. We import that function using the following line:

In [ ]:
from pyhum.preprocessing import clean_text

In [ ]:
def tokenize(text):
    """Transform TEXT into a list of sentences. Lowercase 
    each sentence and remove all punctuation. Finally split each
    sentence into a list of words."""
    # insert your code here

# these tests should return True if your code is correct
print(tokenize("This is a sentence. So, what!") == 
      [["this", "is", "a", "sentence"], ["so", "what"]])

---

## General Text Statistics

> When the next night came, Dinarazad said to her sister Shahrazad: ‘In God’s name, sister, if you are not asleep, then tell us one of your stories!’ Shahrazad answered: ‘With great pleasure! I have heard tell, honoured King, that…’

---

#### Quiz!

In the directory `data/arabian_nights` you will find 999 files. This is because in Burton's translation some nights are missing. The name of the file represents the corresponding night of storytelling in *Alf Laylah Wa Laylah*. Go have a look. Use the tokenize function and the corpus reading function we have defined above and tokenize and clean each night. Store the result in the variable named `corpus`.

In [ ]:
# insert your code here

---

Great job! You now should have a corpus containing 999 texts. It is always important to check whether our code actually produces the desired results. Let's check whether we indeed have 999 texts:

In [ ]:
print(len(corpus))

OK, that seems to be correct. It would be convenient for further processing to have the corpus in chronological order. Let's have a look the first 20 files returned by `list_textfiles`:

In [ ]:
list_textfiles("data/arabian_nights")[:20]

As you can see the files are sorted by their string name and not by their numbering. To be able to sort the files by their numbers we must first remove the extension `.txt` as well as the directory `data/arabian_nights/`.

---

#### Quiz!

**1)** Write a function `remove_txt` that takes as argument a string and some extension that you want to remove. It should return the string without the extension. Tip: use the function `splitext` from the `os.path` module. Look up the documentation [here](http://docs.python.org/3.3/library/os.path.html#os.path.splitext).

In [ ]:
from os.path import splitext

def remove_ext(filename):
    # insert your code here
    
# these tests should return True if your code is correct
print(remove_ext("data/arabian_nights/1.txt") == "data/arabian_nights/1")
print(remove_ext("ridiculous_selfie.jpg") == "ridiculous_selfie")

**2)** Write a function `remove_dir` that takes as argument a filepath and removes the directory from a filepath. Tip: use the function `basename` from the `os.path` module. Look up the document [here](http://docs.python.org/3.3/library/os.path.html#os.path.basename)

In [ ]:
from os.path import basename

def remove_dir(filepath):
    # insert your code here
    
# these tests should return True if your code is correct
print(remove_dir("data/arabian_nights/1.txt") == "1.txt")
print(remove_dir("/a/kind/of/funny/filepath/to/file.txt") == "file.txt")

**3)** Combine the two functions `remove_ext` and `remove_dir` into one function `get_filename`. This function takes as argument a filepath and returns the name (without the extensions) of the file.

In [ ]:
def get_filename(filepath):
    # insert your code here
    
# these tests should return True if your code is correct
print(get_filename("data/arabian_nights/1.txt") == '1')

---

The final step is to convert numbers represented as string (e.g. "1" and "10") to a number. This can be achieved by using the function `int`:

In [ ]:
x_as_string = "1"
x_as_int = int(x_as_string)
print(x_as_int)

The process of converting a string into an integer, is called *type casting*. Strings are different types than integers. To see this, have a look at the following:

In [ ]:
x = "1"
y = "2"
print(x + y)

12? Yes, 12. This is because, as you might remember from the first chapter, we can use the `+` operator to concatenate two strings. If we apply the same operation to integers, as in:

In [ ]:
x = 1
y = 2
print(x + y)

we get the expected result of 3.

---

#### Quiz!

Combine the functions `int` and `get_filename` into the function `get_night` to obtain the integer corresponding to a night.

In [ ]:
def get_night(filepath):
    # insert your code here

# these tests should return True if your code is correct
print(get_night("data/arabian_nights/1.txt") == 1)

---

OK, so now we can convert the filepaths to integers corresponding to the nights of storytelling. But how will we use that to sort the corpus in chronological order? In chapter 1 we briefly discussed how to sort your collection of good reads. In combination with our `get_night` function, we can use `sort` to obtain a nicely chronologically ordered list of stories. Prepare yourself for some real Python magic, because the following lines of code might be a little dazzling...

First we list all files using `list_textfiles` and store it in the variable `filenames`:

In [ ]:
filenames = list_textfiles('data/arabian_nights')

Next we call the function `.sort()` on this list and supply as keyword our function `get_night`:

In [ ]:
filenames.sort(key=get_night)
filenames[:20]

As you can see, we now have a perfectly chronologically ordered list of filenames. But how, **HOW!** did that work? As you might have guessed, the argument of `sort`: `key=get_night`, has something to do with all this magic. Without this argument, Python would just sort the filenames alphabeticaly:

In [ ]:
filenames = list_textfiles('data/arabian_nights')
filenames.sort()
print(filenames[:20])

However, if we supply a function to `key`, Python will internally first apply that function to all items we want to sort. In our case this means Python converts all filepaths to integers. After that Python sorts the list. Then for each converted item it returns the corresponding item in the original list. (Technically this is not an accurate description, but it basically comes down to this.)

If you still feel a little dizzy after all this, don't be afraid. Sometimes it is good enough to use a particular piece of code even if you don't completely understand it. We can now use these functions to reload the corpus, this time in chronological order:

In [ ]:
corpus = []
filenames = list_textfiles("data/arabian_nights")
filenames.sort(key=get_night)
for filename in filenames:
    text = read_file(filename)
    corpus.append(tokenize(text))

---

### Exploratory data analysis

As a first exploratory data analysis, we are going to compute for each night how many sentences it contains and how many words. It is quite easy to count the number of sentences per night, since each night is represented by a list of sentences.

In [ ]:
sentences_per_night = []
for night in corpus:
    sentences_per_night.append(len(night))
print(sentences_per_night[:10])

Using the function `max` we can find out what the highest number of sentences is:

In [ ]:
max(sentences_per_night)

Similarly, if we would like to now what the lowest number of sentences is, we use the function `min`:

In [ ]:
min(sentences_per_night)

---

#### Quiz!

The function `sum` takes a list of numbers as input and returns the sum:

In [ ]:
print(sum([1, 3, 3, 4]))

Use this function to compute the average number of sentences per night. Note if you use Python 2.7, you will need to convert the result of sum, which will be an integer to a `float`, using `float(some_number)`. 

In [ ]:
# if you use Python 3.x, both print statements will return 
# the same thing and you don't need to worry.
number = 1
print(number)
number = float(number)
print(number)

In [ ]:
# insert your code here

---

Given our data structure of a list of sentences which are themselves lists of words, it is a little trickier to count for each night how many words it contains. One possible way is the following:

In [ ]:
words_per_night = []
for night in corpus:
    n_words = 0
    for sentence in night:
        n_words += len(sentence)
    words_per_night.append(n_words)

Make sure you really understand these lines of code as you will need them in the next quiz. 

The suspense created by Shahrazad’s story-telling skills is intriguing, especially the “cliff-hanger” ending each night which she uses to avert her own execution (and possibly that of womanhood). Every night she tells the Sultan a story only to stop at dawn and she picks up the thread the next night. But does it really take the whole night to tell a particular story?

I am not aware of any exact numbers about how many words people speak per minute. Averages seem to fluctuate between 100 and 200 words per minute. Narrators are advised to use approximately 150 words per minute in audiobooks. I suspect that this number is a little lower for live storytelling and assume it lies around 130 words per minute (including pauses). Using this information, we can compute the time it takes to tell a particular story as follows:

$$\textrm{story time}(\textrm{text}) = \frac{\textrm{number of words in text}}{\textrm{number of words per minute}}$$

---

#### Quiz!

**1)** Write a function called `story_time` that takes as input a text. Given a speed of 130 words per minute, compute how long it takes to tell that text.

In [ ]:
def story_time(text):
    # insert your code here

# these tests should return True if your code is correct
print(story_time([["story", "story"]]) * 130 == 2.0)

**2)** Compute the story_time for each night in our corpus. Assign the result to the variable `story_time_per_night`.

In [ ]:
story_time_per_night = []
# insert your code here
print(story_time_per_night[:10])

**3**) Compute the average, minimum and maximum story telling time.

In [ ]:
# insert your code here

---

### Visualizing general statistics

Now that we have computed a range of general statistics for our corpus, it would be nice to visualize them. Python's plotting library *matplotlib* (see [here](http://matplotlib.org)) allows us to produce all kinds of graphs. We could for example, plot for each story, how many sentences it contains:

In [ ]:
import matplotlib.pyplot as plt

plt.plot(sentences_per_night)

---

#### Quiz!

**1)** Can you do the same for `words_per_night`?

In [ ]:
# insert your code here

**2)** And can you do the same for `story_time_per_night`?

In [ ]:
# insert your code here

**3)** In this final exercise we will put everything together what we have learnt so far. We want you to write a function `positions_of` that returns for a given word all sentence positions in the *Arabian Nights* where that word occurs. We are not interested in the positions relative to a particular night, but only to the corpus as a whole. Use that function to find all occurences of the name Sharahzad and store the corresponding indexes in the variable `positions_of_shahrazad`. Do the same thing for the name *Ali*. Store the result in `positions_of_ali`. Finally, find all occurences of *Egypt* and store the indexes in `positions_of_egypt`. Tip: (1) remember that we lowercased the entire corpus! (2) remember that indexes start at 0.

In [ ]:
def positions_of(word):
    #insert your code here

positions_of_shahrazad = positions_of("shahrazad")
positions_of_ali = positions_of("ali")
positions_of_egypt = positions_of("egypt")

If everything went well, the following lines of code should produce a nice dispersion plot of all sentence occurences of Shahrazad, Ali and Egypt in the corpus.

In [ ]:
plt.figure(figsize=(20, 8))
names = ["Shahrazad", "Ali", "Egypt"]
plt.plot(positions_of_shahrazad, [1]*len(positions_of_shahrazad), "|", markersize=100)
plt.plot(positions_of_ali, [2]*len(positions_of_ali), "|", markersize=100)
plt.plot(positions_of_egypt, [0]*len(positions_of_egypt), "|", markersize=100)
plt.yticks(range(len(names)), names)
_ = plt.ylim(-1, 3)

---

> Then Shahrazad reached the morning, and fell silent in the telling of her tale…

---

Ignore the following, it's just here to make the page pretty:

In [2]:
from IPython.core.display import HTML
def css_styling():
    styles = open("styles/custom.css", "r").read()
    return HTML(styles)
css_styling()

---

<p><small><a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/"><img alt="Creative Commons License" style="border-width:0" src="https://i.creativecommons.org/l/by-sa/4.0/88x31.png" /></a><br /><span xmlns:dct="http://purl.org/dc/terms/" property="dct:title">Python Programming for the Humanities</span> by <a xmlns:cc="http://creativecommons.org/ns#" href="http://fbkarsdorp.github.io/python-course" property="cc:attributionName" rel="cc:attributionURL">http://fbkarsdorp.github.io/python-course</a> is licensed under a <a rel="license" href="http://creativecommons.org/licenses/by-sa/4.0/">Creative Commons Attribution-ShareAlike 4.0 International License</a>. Based on a work at <a xmlns:dct="http://purl.org/dc/terms/" href="https://github.com/fbkarsdorp/python-course" rel="dct:source">https://github.com/fbkarsdorp/python-course</a>.</small></p>